sdks/python/apache_beam/testing/synthetic_pipeline.py

```diff
@@ -250,6 +252,95 @@ def default_output_coder(self):
         [beam.coders.BytesCoder(), beam.coders.BytesCoder()])
 
 
+class SyntheticSDFSourceRestrictionProvider(RestrictionProvider):
+  """A `RestrictionProvider` for SyntheticSDFAsSource.
+
+  In initial_restriction(element) and split(element), element means source
+  description.
+  A typical element is like:
+
+    {
+      'key_size': 1,
+      'value_size': 1,
+      'initial_splitting_num_bundles': 2,
+      'initial_splitting_desired_bundle_size': 2,
+      'sleep_per_input_record_sec': 0,
+      'initial_splitting' : 'const'
+
+    }
+
+  """
+
+  def initial_restriction(self, element):
+    return (0, element['num_records'])
+
+  def create_tracker(self, restriction):
+    return restriction_trackers.OffsetRestrictionTracker(
+        restriction[0], restriction[1])
+
+  def split(self, element, restriction):
```

Comment:
Also for my edification, can `split` be called on restrictions that have already been split? If so, would the implementation of this method change? (because it seems that it calculates splits based on the 'global/unsplit' info coming from the element).

TBH you know this way better than me, so I may be barking at the wrong tree : )